### DISCUSSION TITLE
Marker interface for generic math interfaces

### DISCUSSION THREAD

**Discussion by SmartK8** [dotnet/runtime#70203] [Ideas] [answered]
I'm trying to practically use these interfaces. They're great, but it would be even better, if generic math interface had a **marker interface**. Example:

```
// marker interface only
public interface ISignedNumber {}

// standard generic math interface
public interface ISignedNumber<TSelf> : ISignedNumber, INumber<TSelf>...
```

**Advantages**:

- you can use them as a generic constraint without specifying exact generic type
- you can test for the interface presence easier
- it's easy to implement

**Disadvantages**:

- it's just a marker interface
- maybe someone will explain to me: why not

**My logic behind it**: 

These interfaces are to be used in generic classes as constraints. The marker interfaces broaden the ability to use them in these scenarios. You can specify that only signed numbers (for example) will be used and specify the exact type down the line of inheritance.
Attribute would not fulfill the same function. Unless.. attributes as generic constraints.. but let's not open that can of worms.

**Comment by svick** [comment]
> You can specify that only signed numbers (for example) will be used and specify the exact type down the line of inheritance.

What is the advantage of the constraint in that case?

**Reply by SmartK8** [reply]
You can either handle them by determining generic argument in code, or you can use them in an interface and then make specific constraints later in inherited interfaces/structs/records/classes. You're just marking on the ancestor interface that only these types will flow this way.

**Reply by svick** [reply]
> You're just marking on the ancestor interface that only these types will flow this way.

Yes, but why? In what way is the code with the constraint better? (Maybe some example code would help explain that?)

**Reply by SmartK8** [reply]
```
public interface ICalculator
{
    ISignedNumber Add(ISignedNumber a, ISignedNumber b);
}

public interface ICalculator<TType> : ICalculator where Type : ISignedNumber<TType>
{
    TType Add(TType a, TType b) => a + b;

    ISignedNumber ICalculator.Add(ISignedNumber a, ISignedNumber b) => Add((TType) a, (TType) b);
}

public class Calculator : ICalculator<Int32> {}

ICalculator calculator = new Calculator();
ISignedNumber result = calculator.Add(0, 0); // you can only put constrained type here
```

**Reply by svick** [reply]
The problem with the non-generic `ICalculator` interface is that you can only use it if you know how it's implemented. E.g. `calculator.Add(0.0, 0.0)` would throw. This effectively defeats the primary purpose of an interface. Also, it would be quite inefficient, since it would require boxing.

As such, I don't think interfaces like the non-generic `ISignedNumber` should be added, to discourage types like your non-generic `ICalculator`.

**Reply by SmartK8** [reply]
The problem is... I know how it will be implemented. Of course it will throw, because the functionality is not there. The example is just made up on the spot, but the problem is real. The boxing is in this made up sample. I just wanted to point out the advantage. Reality is much more complex than this. It's just a focus point.

**Comment by bartonjs** [comment, marked as answer]
FWIW:

> AVOID using marker interfaces (interfaces with no members).

Framework Design Guidelines, 3rd edition, section 4.6 (p 105 in the print edition).

So it'd need a whole lot of value to justify going against guidelines.  Because of how generic constraints cascade (not well), I don't think there'll be a whole lot of value here.

**Reply by SmartK8** [reply]
Just think about why these interfaces were implemented to do in the first place. The whole point of these interfaces is to be able to discriminate at compile-time what you can input. What's the point if you have to exactly specify the type. Argument doesn't follow to justify that.

**Reply by DrkWzrd** [reply]
You don't have to specify the type you can do `public TResult Method<TResult, TInput>(TInput input) where TInput : ISignedNumber<TInput>{}`

Here you don't specify any type and the constraint remains.

**Reply by tannergooding** [reply]
Right. Its meant for use with generics (and actually requires generics due to how `static virtual` members work).

If we just passed around a bunch of interfaces we'd be boxing, allocating, and have a number of indirections for every single operation. This would be quite inefficient and wouldn't meet the needs or expectations of the community.

Using constrained generics helps ensure performant code can be generated and that it is usable without knowing a concrete type.

**Reply by SmartK8** [reply]
I understand that, but it only helps in few scenarios. When you want to define your own type or when you go after specific types. The reason it was added was because of generic constraint like Enum, struct or class. It will be used obscurely and people will be still asking for the same think as before. For a generic constraint on a number type. Not one exact, all of them at once.

People are now using various crazy workarounds: Object/dynamic/various interfaces. They will use ISpanFormattable now. Because it's the closest to the purpose of generic constraint on numeric types. It's better than the assortment of previous methods, but still includes DateTime, pointers, Guid and such.

You're basically saying: People would use these non-generic interfaces, and it would lead to a lot of boxing (if implemented improperly) and therefore we won't allow it. But then the people, who wanted these constraints, won't use them at all. They will use ISpanFormattable. But people like me who build abstract frameworks will be forced to do runtime checks. So basically no improvement in this area.

I took the time to scour the GitHub, and you can see usages of these concepts. Generic interfaces are used to define custom units and types. All the projects that are abstract frameworks, usually just define their own number types from the scratch, where they can have the craved INumber interface. They don't care about boxing, before you even emerge from their abstract framework, you will deal with it. Various higher level usages do use struct/IEquatable/IComparable/IConvertible combo or out-right runtime-checks, because they don't care about other people using their code (3rd party).

That's why I said, it's great for what it works for, but it could be better. People who want to use generic interfaces will use those, people who want to build abstraction, would use non-generic ones, and deal with the fallout adequately.

**Reply by tannergooding** [reply]
> You're basically saying: People would use these non-generic interfaces, and it would lead to a lot of boxing (if implemented improperly) and therefore we won't allow it

We're also saying that the feature doesn't work. Such interfaces are pointless because you cannot access the requisite `static virtual` APIs.

We cannot expose some non-generic `INumber` interface because it does nothing and provides nothing. You can take/query/constrain to such a thing but then you cannot do anything with it. You do not have access to operators, or members like `Zero`. You cannot do additional checks to get access to those members either and you cannot call an API that is further constrained.

There is no practical use case for such interfaces due to limitations in the language and type system that exist today. You must use generics and the relevant recursive constraint.

**Reply by tannergooding** [reply]
Of course, this could change in the future, but then a corresponding language feature might exist that alleviates the need to provide a non-generic interface at all.

Until such a feature exists, there is no reason to provide a marker interface, particularly since it enables no scenarios (again this being due to limitations of the runtime and language that may be alleviated in some future version of .NET).

**Reply by SmartK8** [reply]
Remember these are marker interfaces, not non-generic versions of math interfaces. You don't have to do anything with them, you only use them in the low-level abstraction. To define non-generic interface methods that operate with these as I have shown in a simplified example above. People will use these, they'll just replace non-generic math interface with Object or something obscure like the struct/interface combo. So this is already possible, except you don't have compile-time check on a numeric type with these. Another use is as a generic constraint, again on lower levels of abstraction. You then build different branches of higher abstraction that are generic. When the implementation leaves the library, it's already all defined in concrete types. So end-user programmer doesn't have to (and isn't allowed to.. actually) worry about boxing.

**Reply by tannergooding** [reply]
We will not be providing marker interfaces for the reasons already expressed/given.

As indicated above, we will instead look at ways to improve the language and runtime experience around generic types and bridging differing constraints or lack thereof.

**Reply by SmartK8** [reply]
I would agree that language feature would be much better than the marker interfaces. I don't want them per se, I want the functionality of them. My thought was it would be easier to want marker interfaces, because they're less of a change, than whole new language feature. I'm just saying the scenarios for this are there, and people who wanted generic constraints on numeric types won't do with generic math interfaces. But it's a work in progress. I will use ISpanFormattable with runtime checks until then.

**Reply by tannergooding** [reply]
Runtime checks won't really fill the gap here, as indicated. You cannot access the relevant members and so you won't have access to things like `int + int` or `checked(int + int)`, etc from a generic context.

Utilizing the generics is really the right and best path forward. You would simply define:
```csharp
public class Calculator<T> : ICalculator<T>
    where T : INumber<T>
{
    // Now all methods in `Calculator` have access to the functionality of `INumber<T>` including operators, etc
}
```

Other generic users with the same constraint or users with concrete types that match the constraint can then use `Calculator<T>` and everything works as expected. It is also immensely more efficient and extensible.

**Reply by SmartK8** [reply]
Unless you have many of those `ICalculator<T>` implementations, and want to store those somewhere, or work with them in a bulk - in abstract. Then you'll need `ICalculator`, but that interface cannot just do nothing. That would defeat the purpose of doing something in bulk. It has to do all the useful operations of generic interfaces otherwise it's unusable. If you know you'll work with a specific T its OK, if you want to work with many Ts, it starts to be a problem.

**Reply by tannergooding** [reply]
That's the point of generic math, it gives you access to operators generically.

You need one generic implementation that constrains its type parameters to the new generic math interfaces. If you want a method which sums values, you define:
```csharp
public static TResult Sum<T, TResult>(IEnumerable<T> values)
    where T : INumber<T>
    where TResult : INumber<TResult>
{
    TResult result = TResult.Zero;

    foreach (var value in values)
    {
        result += TResult.CreateChecked(value);
    }

    return result;
}
```

If you want a method which averages values:
```csharp
public static TResult Average<T, TResult>(IEnumerable<T> values)
    where T : INumber<T>
    where TResult : INumber<TResult>
{
    TResult sum = Sum<T, TResult>(values);
    return TResult.CreateChecked(sum) / TResult.CreateChecked(values.Count());
}
```

You don't need `n` different implementations. You need one generic implementation and now anyone can call `Sum<int, int>` or `Sum<int, long>`, `Sum<byte, BigInteger>`, or `Sum<float, double>`. All of these are possible with a single method implementation (and if you want `T` in, `T` out; you can do that to, just don't include a `TResult` generic parameter and the code still works).

If you want to build `StandardDeviation` on top of these:
```csharp
public static TResult StandardDeviation<T, TResult>(IEnumerable<T> values)
    where T : INumber<T>
    where TResult : IFloatingPointIeee754<TResult>
{
    TResult standardDeviation = TResult.Zero;

    if (values.Any())
    {
        TResult average = Average<T, TResult>(values);
        TResult sum = Sum<TResult, TResult>(values.Select((value) => {
            var deviation = TResult.CreateSaturating(value) - average;
            return deviation * deviation;
        }));
        standardDeviation = TResult.Sqrt(sum / TResult.CreateSaturating(values.Count() - 1));
    }

    return standardDeviation;
}
```

There are no real restrictions or limitations here. You simply just define your algorithms generically, which wasn't previously possible.

**Reply by SmartK8** [reply]
That works if the specificity of type is determined per method (and its nested method calls). But when it is determined on the whole interface, and needs to be consistent across the methods, because they use properties (which can't have local generic constraints), you'll end up with my scenario. In your case there's a multi-type-use wide interface, in mine there are vertically-leveled branches (i.e. multiple Ts) of inheritance of one-type.

**Reply by tannergooding** [reply]
You can also do this at the type level as well.

This is equally as valid and extends to any number of generic type parameters you want to specify.
```csharp
public class NumberMath<T>
    where T : INumber<T>
{
    T _iCanBeAField;

    T ICanBeAProperty { get; set; }

    public static TResult Sum(IEnumerable<T> values)
    {
        TResult result = TResult.Zero;
    
        foreach (var value in values)
        {
            result += TResult.CreateChecked(value);
        }
    
        return result;
    }

    public T ICanBeAMethod(T value)
    {
        /* code */
    }

    public class ICanBeUsedWithNestedTypes
    {
        T _value;

        // etc
    }
    
    public class ICanBeUsedWithNestedGenericTypes<U>
    {
        // etc
    }
}
```

**Reply by SmartK8** [reply]
You're right. That was the missing part. I was skeptical, but converted some portion of my considerable structure, and it seems to magically work out so far. That set of "static abstract" methods on the generic is surprisingly pulling the weight of whole paradigm change. I'm really impressed that this seemingly small toolset can have such impact. It will take time to convert all, and I hope I won't be stopped by something later down the line, but I'm starting to believe. Thanks, I think I've gained new appreciation for this effective change.
